### ISSUE TITLE
Didn't set model to eval mode when validating in chapter 8

### ISSUE THREAD

**Issue by tianlianghai** [deep-learning-with-pytorch/dlwpt-code#104]
In the book and code when we first met resnet, the model has some batch norm layers.
But in validate function, it didn't call model.eval(), I don't think this is true.

![image](https://user-images.githubusercontent.com/75210582/211145816-33152877-5b52-49cf-bd96-42d7161fd498.png)

as a result, in my experiment, while I call model.eval() in the validating function, the result accuracy is lower and loss is higher
![image](https://user-images.githubusercontent.com/75210582/211145957-9e2ab33e-b492-40af-8ff0-2746d9676def.png)

```
def val(model, val_loader, loss_fn):
    model.eval()
    losses = []
    correct = 0
    total = 0 
    for x, y in val_loader:
        with torch.no_grad():
            x, y = x.to(device), y.to(device)
            
            out = model(x)
            loss = loss_fn(out, y)
            losses.append(loss.item())
            _, pred = torch.max(out, dim=1)
            correct += int(torch.sum(pred == y))
            total += y.shape[0]
    loss = sum(losses) / len(losses)
    acc = correct/total
    print(f"val loss:{loss:.3f}, acc:{acc:.3f}")
    model.train()
```
if I add the model.val() line in the code , the validation accuracy even goes done during training, almost always 50%
![image](https://user-images.githubusercontent.com/75210582/211146102-d40e7a73-5e91-4038-983c-1840474d2bc5.png)

while I don't add the model.eval() line the test accuracy actually reached about 85%.
Shouldn't we set model.eval() in validating?

**Comment by evenbrenden** [comment]
I noticed this too and found https://livebook.manning.com/forum?product=huang&comment=576297. It looks like the model implementation for `NetResDeep` is incorrect, which is masked by not calling `model.eval()`. Fixing the model and calling `model.eval()` before validation should give the true accuracies, which are lower than the ones in the book, at least for that particular model.
